### PR TITLE
feat: add agent skills plugin

### PR DESCRIPTION
**Pull request by notowen333**
## Motivation

The AgentSkills.io specification defines a portable format for packaging reusable agent instructions as "skills" — directories containing a `SKILL.md` file with YAML frontmatter metadata and a markdown body. Supporting this format enables our SDK users to leverage a growing ecosystem of community-authored skills and to package their own domain-specific instructions for reuse across agents.

A key design goal is **progressive disclosure**: skill metadata (name, description) is injected into the system prompt so the model knows what's available, but full instructions are only loaded on demand via a tool call. This keeps the system prompt lightweight when many skills are registered but only a few are activated per conversation.

## Public API Changes

New sub-path export `@strands-agents/sdk/vended-plugins/skills` exposing two main constructs:

### `Skill` — data model and loader

```typescript
import { Skill } from '@strands-agents/sdk/vended-plugins/skills'

// From raw content
const skill = Skill.fromContent(`---
name: my-skill
description: Does something useful
---
# Instructions
Follow these steps...`)

// From a skill directory on disk
const skill = Skill.fromFile('./skills/my-skill')

// Load all skills from a parent directory
const skills = Skill.fromDirectory('./skills/')

// From an HTTPS URL (raw SKILL.md content)
const skill = await Skill.fromUrl('https://raw.githubusercontent.com/org/repo/main/SKILL.md')

// Or construct directly
const skill = new Skill({
  name: 'my-skill',
  description: 'A custom skill',
  instructions: 'Do the thing',
})
```

### `AgentSkillsPlugin` — agent integration

```typescript
import { Agent } from '@strands-agents/sdk'
import { AgentSkillsPlugin } from '@strands-agents/sdk/vended-plugins/skills'

const plugin = new AgentSkillsPlugin({
  skills: [
    './skills/pdf-processing',                          // single skill directory
    './skills/',                                         // parent directory (auto-discovers children)
    'https://example.com/skills/my-skill/SKILL.md',     // HTTPS URL to raw SKILL.md
    mySkillInstance,                                     // Skill instance
  ],
  stateKey: 'my_skills',  // custom key for agent.appState (default: 'agent_skills')
})

const agent = new Agent({ model, plugins: [plugin] })
```

The plugin:
- Registers a `skills` tool that the agent calls to load full instructions for a given skill
- Injects an `<available_skills>` XML block into the system prompt before each invocation (idempotent, uses exact-string state tracking via `agent.appState` to replace the previous injection)
- Tracks activated skills per-agent via `agent.appState`
- Supports runtime skill swapping via `plugin.setAvailableSkills()`

## Use Cases

- **Reusable instruction packages**: Package domain-specific agent behaviors (code review, summarization, PDF processing) as portable skill directories that work across agents and teams
- **Progressive disclosure at scale**: Register dozens of skills without bloating the system prompt — the model only loads instructions for skills it actually needs
- **Runtime skill management**: Swap available skills between invocations based on conversation context or user permissions
- **Remote skill loading**: Load skills directly from HTTPS URLs (e.g. GitHub raw content) without requiring local filesystem access

## Description

New feature

## Related Issues

#774 

## Documentation PR

Will follow up when shapes are finalized

## Testing

unit tests and integ tests

- [x] I ran `npm run check`

## Checklist
- [x] I have read the CONTRIBUTING document
- [x] I have added any necessary tests that prove my fix is effective or my feature works
- [ ] I have updated the documentation accordingly
- [ ] I have added an appropriate example to the documentation to outline the feature, or no new docs are needed
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by agent-of-mkmeral** [comment]
## Skills TS Review — PR #807 in strands-agents/sdk-typescript

**TL;DR**: Excellent implementation. Clean architecture, solid test coverage (72 unit tests), and several design improvements over the Python version. I ran **105 tests** (72 original + 33 adversarial) — all pass. Found **3 actionable issues** (1 bug in Python upstream, 1 medium, 1 low) and documented **8 key behavioral differences** vs Python.

<details>
<summary>📋 Full Review & Adversarial Testing Report</summary>

### What the PR Does

Adds `@strands-agents/sdk/vended-plugins/skills` — a complete AgentSkills.io implementation:

- **`Skill`** class — data model + static loaders (`fromFile`, `fromContent`, `fromDirectory`)
- **`AgentSkillsPlugin`** — Plugin that registers a `skills` tool + injects skill metadata into system prompt
- **Progressive disclosure** — metadata in prompt, full instructions loaded on-demand via tool

**Files:** 7 changed (1999 additions, 2 deletions)
- `skill.ts` (388 lines) — Skill parsing/loading
- `agent-skills.ts` (384 lines) — Plugin implementation  
- `index.ts` (31 lines) — Public exports
- `skill.test.node.ts` (519 lines) — 38 unit tests
- `agent-skills.test.node.ts` (492 lines) — 34 unit tests
- `agent-skills.test.node.ts` (integ, 178 lines) — 4 integration tests
- `package.json` — New sub-path export + `yaml` dependency

---

### 🔍 Adversarial Testing Results

**105 tests total** | **105 passed** | **0 failed**

<details>
<summary>33 adversarial tests (injection, edge cases, parity)</summary>

| Category | Tests | Result |
|----------|-------|--------|
| XSS/injection in descriptions | 2 | ✅ |
| YAML injection attempts | 1 | ✅ |
| Extreme skill names (long, unicode, numeric, boolean) | 6 | ✅ |
| Empty/degenerate content | 2 | ✅ |
| XML injection in skill body | 1 | ✅ |
| YAML colon-quoting fallback | 2 | ✅ |
| System prompt marker injection resistance | 2 | ✅ |
| Rapid skill swapping (10 cycles) | 1 | ✅ |
| Empty plugin (zero skills) | 1 | ✅ |
| Duplicate skill activation (idempotent) | 1 | ✅ |
| Path traversal in tool call | 1 | ✅ |
| Symlinks in skills directory | 1 | ✅ |
| Deeply nested directories (direct children only) | 1 | ✅ |
| Concurrent reads | 1 | ✅ |
| 1MB skill body | 1 | ✅ |
| SKILL.md file path (not directory) | 1 | ✅ |
| Python parity checks | 8 | ✅ |

</details>

<details>
<summary>72 original unit tests</summary>

- `skill.test.node.ts`: 38 tests (constructor, fromContent, fromFile, fromDirectory)
- `agent-skills.test.node.ts`: 34 tests (constructor, plugin interface, prompt injection, tool callback, resource listing, setAvailableSkills)

All passing cleanly.

</details>

---

### 🐛 Findings

<details>
<summary>Finding 1 — BUG in Python upstream (not this PR): _fix_yaml_colons doesn't escape quotes</summary>

**This PR actually FIXES a bug that exists in the Python version.**

The TS `fixYamlColons` correctly escapes backslashes and double quotes before wrapping values:
```typescript
const escaped = value!.replace(/\\/g, '\\\\').replace(/"/g, '\\"')
return `${key}: "${escaped}"`
```

The Python version does NOT:
```python
line = f'{key}: "{value}"'  # No escaping!
```

**Reproduction**: A SKILL.md with `description: Use when: user says "hello"` → Python produces broken YAML `"Use when: user says "hello" "` which fails to parse. TS handles it correctly.

**Impact**: The Python `_fix_yaml_colons` will crash on SKILL.md files whose descriptions contain both colons AND double quotes. Should be fixed upstream in sdk-python.

</details>

<details>
<summary>Finding 2 — Medium: System prompt with pre-existing markers gets clobbered</summary>

**Scenario**: A system prompt that already contains `<!-- strands:agent-skills:start -->` text (e.g., documentation about the plugin) will have that section replaced by the regex.

**Code**: `SKILLS_MARKER_REGEX` matches greedily between start/end markers.

**Impact**: Low in practice — users are unlikely to include these exact HTML comments in their prompts. The Python version has a similar (actually worse) issue with exact-string matching. This is an inherent limitation of marker-based injection.

**Mitigation**: Not blocking for merge.

</details>

<details>
<summary>Finding 3 — Low: New `yaml` production dependency</summary>

The PR adds `yaml@^2.8.3` as a production dependency. This is the [yaml npm package](https://www.npmjs.com/package/yaml) (YAML 1.2 parser/stringifier).

- Size: ~120KB minified
- No transitive dependencies
- Well-maintained (monthly releases)
- Required for SKILL.md frontmatter parsing

**Consideration**: For customers who don't use skills, this is dead weight. But at 120KB with zero transitive deps, it's acceptable. Not blocking.

</details>

---

### 🔄 Python Version Comparison

<details>
<summary>Detailed TS vs Python comparison (10 differences)</summary>

| # | Aspect | TypeScript (this PR) | Python (sdk-python) | Assessment |
|---|--------|---------------------|--------------------|-----------|
| 1 | **Class naming** | `AgentSkillsPlugin` | `AgentSkills` | TS follows `*Plugin` convention ✅ |
| 2 | **Plugin interface** | `implements Plugin` (interface) | `extends Plugin` (base class) | TS uses composition, Python uses inheritance |
| 3 | **Hook registration** | `agent.addHook(BeforeInvocationEvent, ...)` in `initAgent` | `@hook` decorator on `_on_before_invocation` | Python uses decorators for auto-discovery |
| 4 | **Tool registration** | `tool()` factory with Zod schema | `@tool(context=True)` decorator | TS gets runtime validation via Zod ✅ |
| 5 | **State management** | `agent.appState.get('strands:agent-skills:activated')` → flat array | `agent.state.get('agent_skills')['activated_skills']` → nested dict | Different keys AND structure ⚠️ |
| 6 | **Prompt injection** | Deterministic HTML comment markers + regex | Exact-string matching from state | **TS is more robust** ✅ — survives external prompt modifications |
| 7 | **XML escaping** | Escapes `& < > " '` (5 chars) | Escapes `& < >` only (3 chars via `xml.sax.saxutils.escape`) | TS is more thorough ✅ |
| 8 | **Colon-quoting** | Escapes `\` and `"` before wrapping | **Does NOT escape** → bug with values containing `"` | TS fixes a Python bug ✅ |
| 9 | **SystemPrompt types** | Handles `string \| null \| undefined \| SystemContentBlock[]` | Handles `str \| None` only | TS handles more prompt types ✅ |
| 10 | **Skill source types** | `string \| Skill` | `str \| Path \| Skill` | Python supports `Path` objects |

**Key takeaway**: The TS implementation is a clean port with several meaningful improvements:
1. Marker-based prompt injection (vs fragile exact-string matching)
2. Proper quote escaping in YAML fallback
3. Zod runtime validation for tool inputs
4. Support for `SystemContentBlock[]` prompt types

**Cross-SDK state incompatibility** (item 5): If a user ever needs to migrate skill state between TS and Python agents (unlikely but possible in a polyglot system), the different state keys/structure would break. Worth documenting.

</details>

---

### 🎮 DevX Examples

<details>
<summary>How customers will use this</summary>

**Basic usage — load skills from filesystem:**
```typescript
import { Agent } from '@strands-agents/sdk'
import { AgentSkillsPlugin } from '@strands-agents/sdk/vended-plugins/skills'

const plugin = new AgentSkillsPlugin({
  skills: ['./skills/'],  // auto-discovers all SKILL.md subdirs
})

const agent = new Agent({ model, plugins: [plugin] })
await agent.invoke('Activate the code-review skill and review this PR')
```

**Programmatic skill creation:**
```typescript
import { Skill, AgentSkillsPlugin } from '@strands-agents/sdk/vended-plugins/skills'

const skill = new Skill({
  name: 'summarizer',
  description: 'Summarize text into bullet points',
  instructions: '# Steps\n1. Read the text\n2. Extract key points\n3. Format as bullets',
})

const plugin = new AgentSkillsPlugin({ skills: [skill] })
```

**Mixed sources (files + instances):**
```typescript
const plugin = new AgentSkillsPlugin({
  skills: [
    './skills/pdf-processing',     // single skill directory
    './skills/',                    // parent directory (auto-discovers)
    myCustomSkillInstance,          // Skill instance
  ],
})
```

**Runtime skill swapping:**
```typescript
const plugin = new AgentSkillsPlugin({ skills: [skillA, skillB] })
const agent = new Agent({ model, plugins: [plugin] })

await agent.invoke('Use skill A')  // Agent sees skillA, skillB

plugin.setAvailableSkills([skillC])  // Swap!

await agent.invoke('Use skill C')  // Agent now sees only skillC
```

**What the agent sees in system prompt:**
```xml
<available_skills>
<skill>
<name>summarizer</name>
<description>Summarize text into bullet points</description>
<location>/path/to/summarizer/SKILL.md</location>
</skill>
</available_skills>
```

**What happens when agent calls the `skills` tool:**
```
Agent: I'll activate the summarizer skill.
→ skills({ skill_name: "summarizer" })
← "# Steps\n1. Read the text\n2. Extract key points\n3. Format as bullets\n\n---\nAllowed tools: bash, file-editor\nLocation: /path/to/summarizer/SKILL.md"
```

</details>

---

### ✅ What Survived Testing

- **YAML injection** — XSS, colons, quotes, numeric/boolean names all handled
- **System prompt injection** — markers prevent duplication, survive external modifications
- **Path traversal** — `../../../etc/passwd` in skill_name returns "not found" (safe)
- **Large payloads** — 1MB skill body parses fine
- **Idempotency** — multiple prompt injections produce exactly one `<available_skills>` block
- **Rapid swapping** — 10 consecutive skill set changes work perfectly
- **Edge cases** — symlinks, nested dirs, concurrent reads, empty bodies all handled
- **Type safety** — `string | null | undefined | SystemContentBlock[]` prompt types all supported

</details>

---

### Verdict

The PR is **ready to merge**. The implementation is cleaner than the Python version in several ways (marker-based injection, Zod validation, proper YAML escaping). The code is well-structured, extensively tested, and follows SDK conventions.

**Before merge, consider:**
1. ✅ Fix the Python `_fix_yaml_colons` bug upstream (found via this review)
2. 📝 AGENTS.md update (add `src/vended-plugins/` to directory structure)
3. 📝 Documentation PR (stated as follow-up — that's fine)

**Nice to have (not blocking):**
- Document the state key differences vs Python for cross-SDK compatibility
- Consider whether `AgentSkillsPlugin` vs Python's `AgentSkills` naming difference is intentional

---
🤖 *AI agent review — 105 tests (72 unit + 33 adversarial), all passing. [Strands Agents](https://github.com/strands-agents). Feedback welcome!*

File: src/vended-plugins/skills/agent-skills.ts [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+  /** Maximum number of resource files to list in skill responses. Defaults to 20. */
+  maxResourceFiles?: number | undefined
+
+  /** If true, throw on skill validation issues. If false (default), warn and load anyway. */
+  strict?: boolean | undefined
+}
+
+const SKILLS_MARKER_START = '<!-- strands:agent-skills:start -->'
+const SKILLS_MARKER_END = '<!-- strands:agent-skills:end -->'
+const SKILLS_MARKER_REGEX = new RegExp(`${escapeRegex(SKILLS_MARKER_START)}[\\s\\S]*?${escapeRegex(SKILLS_MARKER_END)}`)
```
Comment (mkmeral) [inline src/vended-plugins/skills/agent-skills.ts]: should we do regex? I'm not a fan. In python, we keep track of the injected string in agent state.

File: src/vended-plugins/skills/agent-skills.ts [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+ * Escape special regex characters in a string.
+ */
+function escapeRegex(str: string): string {
+  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
+}
+
+/**
+ * Escape XML special characters in text content.
+ */
+function escapeXml(text: string): string {
```
Comment (mkmeral) [inline src/vended-plugins/skills/agent-skills.ts:53]: any libraries we can use instead, or is this the best approach?

File: src/vended-plugins/skills/agent-skills.ts (hunk 1 above)
Comment (notowen333) [reply src/vended-plugins/skills/agent-skills.ts]: I think it comes down to

- Exact string replace is simpler and more deterministic when you fully control prompt contents.
- Regex sentinel is better when the prompt may evolve or be touched by other code.

I don't feel strongly but the sentinel seemed more robust to me.

File: src/vended-plugins/skills/agent-skills.ts (hunk 2 above)
Comment (notowen333) [reply src/vended-plugins/skills/agent-skills.ts:53]: I looked into a few

[he](https://www.npmjs.com/package/he) no publish since 2018
[lodash](https://www.npmjs.com/package/lodash) html focused

since this is so small; a few lines helper is acceptable I think

File: src/vended-plugins/skills/skill.ts [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+ * const skill = Skill.fromFile('./skills/my-skill')
+ *
+ * // From raw SKILL.md content
+ * const skill = Skill.fromContent('---\nname: my-skill\n...')
+ *
+ * // Load all skills from a parent directory
+ * const skills = Skill.fromDirectory('./skills/')
+ * ```
+ */
+export class Skill {
```
[the remark this answers is not in the document]
Comment (notowen333) [reply src/vended-plugins/skills/skill.ts:232]: Other data classes in the codebase like `TextBlock` and `Message` use the public `readonly` pattern for immutable fields. The _ getter pattern is a good fit for the `AgentSkillsPlugin` which has mutable internal state

File: src/vended-plugins/skills/skill.ts [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+    for (const [k, v] of Object.entries(metadataRaw)) {
+      metadata[String(k)] = v
+    }
+  }
+
+  const skillLicense = frontmatter['license']
+  const compatibility = frontmatter['compatibility']
+
+  return new Skill({
+    name: frontmatter['name'] as string,
```
[the remark this answers is not in the document]
Comment (notowen333) [reply src/vended-plugins/skills/skill.ts:196]: buildSkillFromFrontmatter is a private module-level function (not exported, not a class method) with exactly one call site in fromContent() at skill.ts:343

File: src/vended-plugins/skills/agent-skills.ts [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+        }
+        return this._activateSkill(input.skill_name, context)
+      },
+    })
+  }
+
+  /**
+   * Handle skill activation from the tool callback.
+   */
+  private _activateSkill(skillName: string, context: ToolContext): string {
```
[the remark this answers is not in the document]
Comment (notowen333) [reply src/vended-plugins/skills/agent-skills.ts:295]: ToolContext is a typed interface — if context is non-null, context.agent is guaranteed by the type system.

File: src/vended-plugins/skills/agent-skills.ts [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+    } else if (typeof systemPrompt === 'string') {
+      // Strip existing markers, then append fresh block
+      const stripped = systemPrompt.replace(SKILLS_MARKER_REGEX, '').trimEnd()
+      agent.systemPrompt = stripped ? `${stripped}\n\n${wrappedBlock}` : wrappedBlock
+    } else {
+      // SystemContentBlock[] — filter out the TextBlock containing our marker, append new one
+      const filtered: SystemContentBlock[] = systemPrompt.filter(
+        (block) => !(block.type === 'textBlock' && block.text.includes(SKILLS_MARKER_START))
+      )
+      filtered.push(new TextBlock(wrappedBlock))
```
[the remark this answers is not in the document]
Comment (notowen333) [reply src/vended-plugins/skills/agent-skills.ts]: The TextBlock constructor takes a plain string (messages.ts:158), so new TextBlock(wrappedBlock) at agent-skills.ts:277 is correct

File: src/vended-plugins/skills/agent-skills.ts [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+    const files: string[] = []
+
+    for (const dirName of RESOURCE_DIRS) {
+      const resourceDir = join(skillPath, dirName)
+      if (!existsSync(resourceDir) || !statSync(resourceDir).isDirectory()) {
+        continue
+      }
+
+      const entries = readdirSync(resourceDir, { recursive: true, encoding: 'utf-8' })
+      for (const entry of [...entries].sort()) {
```
[the remark this answers is not in the document]
Comment (notowen333) [reply src/vended-plugins/skills/agent-skills.ts]: fixing

File: src/vended-plugins/skills/skill.ts [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+ */
+function fixYamlColons(yamlStr: string): string {
+  return yamlStr
+    .split('\n')
+    .map((line) => {
+      const match = line.match(/^(\s*\w[\w-]*):\s+(.+)$/)
+      if (match) {
+        const key = match[1]
+        const value = match[2]
+        if (value!.includes(':') && !value!.startsWith('"') && !value!.startsWith("'")) {
```
[the remark this answers is not in the document]
Comment (notowen333) [reply src/vended-plugins/skills/skill.ts]: fixed

File: package.json [hunk 9]
```diff
@@ -157,7 +161,8 @@
   "dependencies": {
     "@aws-sdk/client-bedrock-runtime": "^3.943.0",
     "@types/json-schema": "^7.0.15",
-    "uuid": "^13.0.0"
+    "uuid": "^13.0.0",
+    "yaml": "^2.8.3"
```
[the remark this answers is not in the document]
Comment (notowen333) [reply package.json:166]: not an issue `<3` is the same

[the remark this answers is not in the document]
Comment (notowen333) [reply package.json:166]: A tiny utility package for a good DX is worth it

File: src/vended-plugins/skills/agent-skills.ts [hunk 10]
```diff
@@ -0,0 +1,422 @@
...
+    if (systemPrompt == null || typeof systemPrompt === 'string') {
+      let currentPrompt = systemPrompt ?? ''
+
+      // Remove previously injected XML by exact match
+      const lastInjectedXml = this._getStateField(agent, 'lastInjectedXml') as string | undefined
+      if (lastInjectedXml != null) {
+        if (currentPrompt.includes(lastInjectedXml)) {
+          currentPrompt = currentPrompt.replace(lastInjectedXml, '')
+        } else {
+          logger.warn('unable to find previously injected skills XML in system prompt, re-appending')
```
[the remark this answers is not in the document]
Comment (notowen333) [reply src/vended-plugins/skills/agent-skills.ts:361]: yes this is intended and the same behavior as python

File: src/vended-plugins/skills/agent-skills.ts [hunk 11]
```diff
@@ -0,0 +1,422 @@
...
+  skills: SkillSource[]
+
+  /** Maximum number of resource files to list in skill responses. Defaults to 20. */
+  maxResourceFiles?: number | undefined
+
+  /** If true, throw on skill validation issues. If false (default), warn and load anyway. */
+  strict?: boolean | undefined
+}
+
+const STATE_KEY = 'strands:agent-skills'
```
Comment (mkmeral) [inline src/vended-plugins/skills/agent-skills.ts]: should we keep the same key as python, for cross-language sessions? 🤔   https://github.com/strands-agents/sdk-python/blob/main/src/strands/vended_plugins/skills/agent_skills.py#L26

File: src/vended-plugins/skills/agent-skills.ts [hunk 12]
```diff
@@ -0,0 +1,422 @@
...
+          // Try loading as a single skill (directory with SKILL.md or SKILL.md file)
+          const skill = Skill.fromFile(source, { strict: this._strict })
+          if (resolved.has(skill.name)) {
+            logger.warn(`name=<${skill.name}> | duplicate skill name, overwriting previous skill`)
+          }
+          resolved.set(skill.name, skill)
+        } catch (singleSkillError) {
+          // Fall back to loading as a parent directory containing skill subdirectories
+          logger.debug(`path=<${source}> | not a single skill (${singleSkillError}), trying as directory`)
+          try {
```
Comment (mkmeral) [inline src/vended-plugins/skills/agent-skills.ts:238]: code nit (not blocking): not a fan of nested try catches

**Comment by notowen333** [comment]
The newest commit uses the same key in AgentState for the skills XML and uses

It also achieves parity with pulling skills from HTTPS URLs which was introduced in python recently. https://github.com/strands-agents/sdk-python/pull/2091

And noting the divergences here based on async JS behavior: 

```
Divergences from Python (all stem from fetch being async in JS):
  - fromUrl is async — no synchronous HTTP client in Node.js
  - _resolveSkills returns { skills, ready } so the constructor remains sync while URL fetches
  complete in the background; initAgent awaits ready before registering hooks
  - Uses globalThis.fetch + AbortSignal.timeout(30_000) instead of urllib (standard Node.js 18+
  built-in, already used in the codebase)
  - Error instead of RuntimeError/ValueError (TS convention — single Error class)
```

File: src/vended-plugins/skills/skill.ts [hunk 13]
```diff
@@ -0,0 +1,437 @@
...
+      throw new Error(`url=<${url}> | not a valid HTTPS URL`)
+    }
+
+    logger.info(`url=<${url}> | fetching skill content`)
+
+    let content: string
+    try {
+      const response = await globalThis.fetch(url, {
+        headers: { 'User-Agent': 'strands-agents-sdk' },
+        signal: AbortSignal.timeout(30_000),
```
[the remark this answers is not in the document]
Comment (notowen333) [reply src/vended-plugins/skills/skill.ts]: this is parity with python but will move to a constant

File: src/vended-plugins/skills/agent-skills.ts [hunk 14]
```diff
@@ -0,0 +1,473 @@
...
+   * Each source can be a Skill instance, a path to a skill directory,
+   * a path to a parent directory containing multiple skills, or an
+   * HTTPS URL pointing to a SKILL.md file.
+   *
+   * Synchronous sources (Skill instances and filesystem paths) are resolved
+   * immediately into the returned map. Async sources (URLs) are resolved in
+   * the background; the returned `ready` promise resolves when all URL
+   * fetches have completed and the map has been updated.
+   */
+  private _resolveSkills(sources: SkillSource[]): { skills: Map<string, Skill>; ready: Promise<void> } {
```
[the remark this answers is not in the document]
Comment (notowen333) [reply src/vended-plugins/skills/agent-skills.ts:185]: fixed with `setAvailableSkills`

File: src/vended-plugins/skills/agent-skills.ts (hunk 11 above)
Comment (notowen333) [reply src/vended-plugins/skills/agent-skills.ts]: synced with the team and decided even if it breaks TS naming convention we'll take the existing name for backwards compatibility. also added the same advice in `AGENTS.md`

File: src/vended-plugins/skills/agent-skills.ts (hunk 12 above)
Comment (notowen333) [reply src/vended-plugins/skills/agent-skills.ts:238]: fixed

File: src/vended-plugins/skills/agent-skills.ts [hunk 15]
```diff
@@ -0,0 +1,492 @@
...
+   * Returns the skills activation tool for auto-registration with the agent.
+   */
+  getTools(): Tool[] {
+    return [this._createSkillsTool()]
+  }
+
+  /**
+   * Get the list of available skills.
+   */
+  getAvailableSkills(): readonly Skill[] {
```
Comment (mkmeral) [inline src/vended-plugins/skills/agent-skills.ts]: should this wait for `_ready`? The main reason we are populating the skills in init is so that it can be modified/updated before applying to the agent, programmatically. If I "init skills with urls, filter the skills (for security? idk), create the agent", my filter logic can actually miss some URL based skills. We should either expose ready, or make getAvailableSkills wait ready too

Comment (notowen333) [reply src/vended-plugins/skills/agent-skills.ts]: added `await this._ready`

**Review by mkmeral — APPROVED** [review]
